### ISSUE TITLE
Files not being saved on exit

### ISSUE THREAD

**Issue by rlabbe** [spyder-ide/spyder#17479]
### What steps will reproduce the problem?

Open spyder, create a new file, type some random characters in it. Try to close spyder using the window's 'x' icon. Spyder prompts you to save the file with the name 'untitled0.py'. Choose a directory to save to, press okay in the save dialog and spyder instantly shuts down.

However, the file is not saved anywhere - if you reopen spyder untitled0.py is not reopened, and a hard drive search confirms it was not saved anywhere.

This time create a new file, type some random characters, and choose file->save. Save it someplace. Using file explorer confirm that it did save. Now, type some more characters in the file, the press the window's 'x' to shut down spyder. Spyder prompts you to save the file. Choose to save it. Again, spyder seems to instantly shut down. If you restart spyder or inspect the file with an editor you will see that the changes made to the file that you made after the first save were not saved on shutdown. 

edit: I submitted via spyder, I thought it would copy version information. I am on Windows 11, Spyder 5.1.5, python 3.9.7 64 bit, using the Anaconda distribution. When I copy the version information from the About box I get the following, note that I am NOT on Windows 10 but 11. 

* Spyder version: 5.1.5 None
* Python version: 3.9.7 64-bit
* Qt version: 5.9.7
* PyQt5 version: 5.9.2
* Operating System: Windows 10
*

**Comment by dalthviz** [comment]
Hi @rlabbe thank you for the feedback! I was unable to reproduce this (however I'm using Spyder 5.2.2 on Windows 10) Could you try with Spyder 5.2.2?

For that, if you want to do it through your anaconda installation, you will need to use the conda-forge channel, so creating a new env could be better to prevent damaging your current setup. YOu can do it by running from an anaconda prompt something like:

```
conda create -n spyder-env -c conda-forge python=3.9.7 spyder=5.2.2
```
You can also use our Windows installer: https://github.com/spyder-ide/spyder/releases/download/v5.2.2/Spyder_64bit_full.exe (for more info regarding the installers and how to use them with packages outside of the ones eing bundled with them you can check the release page: https://github.com/spyder-ide/spyder/releases/latest)

**Comment by rlabbe** [comment]
I put some logging.error messages in the save path of editor.py and tracked down the problem. I have auto format files on save turned on. That puts you in this path in save():

            if self.format_on_save and finfo.editor.formatting_enabled:
                # Autoformat document and then save
                finfo.editor.sig_stop_operation_in_progress.connect(
                    functools.partial(self._save_file, finfo, index))
                finfo.editor.format_document()
            else:
                self._save_file(finfo, index)

I don't understand the sig_stop_operation_in_progress code, but it doesn't appear to be saving the file. With format on save turned off the else path is taken and the file is properly saved. If you can't duplicate the problem by turning on the auto format option, I can put some more logging calls in to narrow down the problem further for you, but I might need some guidance because like I said I don't understand how that code causes save to be called, or if there is just a missing call to save in this control path.

This is still in the 5.1.5 release; I'll try 5.2.2 in a moment

**Comment by dalthviz** [comment]
Thank you for checking further into this and the new info @rlabbe ! I was able to reproduce this when setting the autoformat option on :+1:

We will try to fix this in a future release, but also if you want to give us a hand submitting a fix for this let us know!

Just in case pinging @andfoy and @steff456 (maybe they know how the fix should look)

**Comment by rlabbe** [comment]
@dalthviz I don't mind contributing, but I absolutely don't understand the code with the functools.partial. To me it looks like the else should be removed and the save unindented one level so that it always happens, but who knows. If so, it is a 2 minute fix that anyone on the team could do without me.

**Comment by dalthviz** [comment]
No problem @rlabbe I'm not sure about the fix either so that's why I pinged more people to check this. Anyhow, thank you for the feedback and the info to isolate the issue to a specific part of the code :+1:

**Comment by ccordoba12** [comment]
Hey @rlabbe, this was solved by @dalthviz in pull request #17854, which will be part of our next version (5.3.1), to be released at the end of the week.
